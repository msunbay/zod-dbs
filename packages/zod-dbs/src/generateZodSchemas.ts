import { logDebug } from 'zod-dbs-core';

import type {
  ZodDbsConfig,
  ZodDbsProvider,
  ZodDbsRenderer,
  ZodDbsSchemaInfo,
  ZodDbsZodVersion,
} from 'zod-dbs-core';

import { DEFAULT_CONFIGURATION, DEFAULT_OUTPUT_DIR } from './constants.js';
import { generateConstantsFile } from './generate/generateConstantsFile.js';
import { generateIndexFiles } from './generate/generateIndexFiles.js';
import { generateSchemaFiles } from './generate/generateSchemaFile.js';
import { generateTypesFile } from './generate/generateTypesFile.js';
import { Zod3Renderer } from './renderers/Zod3Renderer.js';
import { Zod4MiniRenderer } from './renderers/Zod4MiniRenderer.js';
import { Zod4Renderer } from './renderers/Zod4Renderer.js';
import { ZodBaseRenderer } from './renderers/ZodBaseRenderer.js';
import { clearTablesDirectory, ensureFolder } from './utils/index.js';

export interface ZodDbsGenerateOptions {
  provider: ZodDbsProvider;
  renderer?: ZodDbsRenderer;
  config: ZodDbsConfig;
}

const createRenderer = (zodVersion: ZodDbsZodVersion | undefined) => {
  if (zodVersion === '3') {
    return new Zod3Renderer();
  }

  if (zodVersion === '4') {
    return new Zod4Renderer();
  }

  if (zodVersion === '4-mini') {
    return new Zod4MiniRenderer();
  }

  return new ZodBaseRenderer();
};

/**
 * Generates Zod schemas for all tables in the specified database schema.
 */
export const generateZodSchemas = async ({
  provider,
  renderer,
  config,
}: ZodDbsGenerateOptions): Promise<ZodDbsSchemaInfo> => {
  const generateConfig = {
    ...DEFAULT_CONFIGURATION,
    ...config,
  };

  const {
    outputDir = DEFAULT_OUTPUT_DIR,
    cleanOutput,
    onProgress,
    zodVersion,
  } = generateConfig;

  if (cleanOutput) {
    clearTablesDirectory(outputDir);
  }

  await ensureFolder(outputDir);

  logDebug(`Connecting to database`);

  const schema = await provider.getSchemaInformation(config);

  onProgress?.('generating', { total: schema.tables.length });

  logDebug(`Generating zod schemas for ${schema.tables.length} tables'`);

  const schemaRenderer = renderer ?? createRenderer(zodVersion);

  for (const table of schema.tables) {
    await generateSchemaFiles(table, schemaRenderer, generateConfig);
  }

  await generateIndexFiles(schema, generateConfig);
  await generateConstantsFile(schema, generateConfig);
  await generateTypesFile(schema, generateConfig);

  onProgress?.('done');

  return schema;
};
