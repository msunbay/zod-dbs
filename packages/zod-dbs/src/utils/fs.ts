import { existsSync, readdirSync, unlinkSync } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import path from 'node:path';
import { logDebug } from 'zod-dbs-core';

import type { ZodDbsTableType } from 'zod-dbs-core';

/**
 * Deletes all .ts files in the output schemas folder.
 */
export function clearTablesDirectory(outputPath: string) {
  // Check if the directory exists first
  if (!existsSync(outputPath)) {
    logDebug(`Directory ${outputPath} does not exist, nothing to clear`);
    return;
  }

  const files = readdirSync(outputPath, {
    recursive: true,
    withFileTypes: true,
  });

  for (const file of files) {
    if (file.isDirectory()) continue;
    if (!file.name.endsWith('.ts')) continue;

    const filePath = path.join(file.path, file.name);
    unlinkSync(filePath);
  }

  if (files.length > 0) {
    logDebug(`Deleted all .ts files in ${outputPath}`);
  }
}

export async function ensureFolder(folderPath: string) {
  const exists = existsSync(folderPath);

  if (!exists) {
    await mkdir(folderPath, { recursive: true });
  }
}

export const getOutputFolder = (type: ZodDbsTableType): string => {
  switch (type) {
    case 'table':
      return 'tables';
    case 'materialized_view':
      return 'materialized_views';
    case 'view':
      return 'views';
    case 'foreign_table':
      return 'foreign_tables';
    default:
      return 'others';
  }
};
