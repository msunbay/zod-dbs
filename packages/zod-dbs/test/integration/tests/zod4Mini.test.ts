import fs from 'node:fs/promises';
import path from 'node:path';

import { generateZodSchemas } from '../../../src/generateZodSchemas.js';
import {
  createTestProvider,
  getOutputDir,
  getOutputFiles,
} from '../testDbUtils.js';

const provider = createTestProvider();

it('generates schemas compatible with zod version 4 mini', async () => {
  const outputDir = getOutputDir('zod4Mini', 'default');

  await generateZodSchemas({
    provider,
    config: {
      outputDir,
      moduleResolution: 'esm',
      zodVersion: '4-mini',
    },
  });

  const outputFiles = await getOutputFiles(outputDir);

  for (const file of outputFiles) {
    const content = await fs.readFile(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});

it('renders without case transforms when "caseTransform" is false', async () => {
  const outputDir = getOutputDir('zod4Mini', 'noCaseTransform');

  await generateZodSchemas({
    provider,
    config: {
      outputDir,
      moduleResolution: 'esm',
      zodVersion: '4-mini',
      caseTransform: false,
    },
  });

  const outputFiles = await getOutputFiles(outputDir);

  for (const file of outputFiles) {
    const content = await fs.readFile(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});
