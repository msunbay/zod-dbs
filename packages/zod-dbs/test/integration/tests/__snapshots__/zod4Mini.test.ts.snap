// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates schemas compatible with zod version 4 mini > constants.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export const TABLE_USERS = 'users';
"
`;

exports[`generates schemas compatible with zod version 4 mini > tables/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './users/index.js';
"
`;

exports[`generates schemas compatible with zod version 4 mini > tables/users/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`generates schemas compatible with zod version 4 mini > tables/users/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod/mini';

export const USER_ROLES = ['admin', 'editor', 'viewer'] as const;

/**
 * Base read schema for the "public.users" table.
 * Validates raw rows read from the database (no casing transforms applied yet).
 */
export const UsersTableBaseSchema = z.object({
     /**
      * dataType: int4
      * defaultValue: nextval('users_id_seq'::regclass)
      */
    id: z.number(),
     /**
      * dataType: varchar
      */
    name: z.string(),
     /**
      * dataType: varchar
      */
    email: z.pipe(z.optional(z.nullable(z.string())), z.transform(val => val ?? undefined)),
     /**
      * dataType: varchar
      * defaultValue: 'active'::character varying
      */
    status: z.pipe(z.optional(z.nullable(z.string())), z.transform(val => val ?? undefined)),
     /**
      * dataType: jsonb
      */
    profile: z.pipe(z.optional(z.nullable(z.json())), z.transform(val => val ?? undefined)),
     /**
      * dataType: timestamptz
      * defaultValue: now()
      */
    created_at: z.pipe(z.optional(z.nullable(z.coerce.date())), z.transform(val => val ?? undefined)),
     /**
      * dataType: _text
      */
    roles: z.pipe(z.optional(z.nullable(z.array(z.enum(USER_ROLES)))), z.transform(val => val ?? undefined)),
     /**
      * dataType: _timestamp
      */
    dates: z.pipe(z.optional(z.nullable(z.array(z.coerce.date()))), z.transform(val => val ?? undefined)),
});

/**
 * The base record type for the "public.users" table.
 * This type represents the raw database record without case transforms.
 */
export type UserBaseRecord = z.output<typeof UsersTableBaseSchema>;

/**
 * Read transform for the "public.users" table.
 * Maps database fields to app case properties.
 */
export const transformUserBaseRecord = (data: UserBaseRecord): {
    id: UserBaseRecord['id'],
    name: UserBaseRecord['name'],
    email?: UserBaseRecord['email'],
    status?: UserBaseRecord['status'],
    profile?: UserBaseRecord['profile'],
    createdAt?: UserBaseRecord['created_at'],
    roles?: UserBaseRecord['roles'],
    dates?: UserBaseRecord['dates'],
} => ({
    id: data.id,
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    createdAt: data.created_at,
    roles: data.roles,
    dates: data.dates,
});

/**
 * Read schema for the "public.users" table (after casing transform).
 */
export const UsersTableSchema = z.pipe(UsersTableBaseSchema, z.transform(transformUserBaseRecord));

/**
 * Base insert/write schema for the "public.users" table (no casing transforms).
 */
export const UsersTableInsertBaseSchema = z.object({
     /**
      * dataType: varchar
      */
    name: z.string().check(z.maxLength(100)),
     /**
      * dataType: varchar
      */
    email: z.optional(z.nullable(z.string().check(z.maxLength(255)))),
     /**
      * dataType: varchar
      * defaultValue: 'active'::character varying
      */
    status: z.optional(z.nullable(z.string().check(z.maxLength(20)))),
     /**
      * dataType: jsonb
      */
    profile: z.pipe(z.optional(z.nullable(z.json())), z.transform((value) => value ? JSON.stringify(value) : value)),
     /**
      * dataType: timestamptz
      * defaultValue: now()
      */
    createdAt: z.optional(z.nullable(z.date())),
     /**
      * dataType: _text
      */
    roles: z.optional(z.nullable(z.array(z.enum(USER_ROLES)))),
     /**
      * dataType: _timestamp
      */
    dates: z.optional(z.nullable(z.array(z.date()))),
});

/**
 * The base record type for the "public.users" table.
 * This type represents an insertable database record before casing transformations are applied.
 */
export type UserInsertBaseRecord = z.output<typeof UsersTableInsertBaseSchema>;

/**
 * The base record type for the "public.users" table.
 * This type represents an updatable database record before casing transformations are applied.
 */
export type UserUpdateBaseRecord = Partial<UserInsertBaseRecord>;

/**
 * Insert transform for the "public.users" table.
 * Maps app cased properties to database fields.
 */
export const transformUserInsertBaseRecord = (data: UserInsertBaseRecord): {
    name: UserInsertBaseRecord['name'],
    email?: UserInsertBaseRecord['email'],
    status?: UserInsertBaseRecord['status'],
    profile?: UserInsertBaseRecord['profile'],
    created_at?: UserInsertBaseRecord['createdAt'],
    roles?: UserInsertBaseRecord['roles'],
    dates?: UserInsertBaseRecord['dates'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
    roles: data.roles,
    dates: data.dates,
});

/**
 * Update transform for the "public.users" table.
 * Maps app cased properties to database fields.
 */
export const transformUserUpdateBaseRecord = (data: UserUpdateBaseRecord): {
    name: UserUpdateBaseRecord['name'],
    email?: UserUpdateBaseRecord['email'],
    status?: UserUpdateBaseRecord['status'],
    profile?: UserUpdateBaseRecord['profile'],
    created_at?: UserUpdateBaseRecord['createdAt'],
    roles?: UserUpdateBaseRecord['roles'],
    dates?: UserUpdateBaseRecord['dates'],
} => ({
    name: data.name,
    email: data.email,
    status: data.status,
    profile: data.profile,
    created_at: data.createdAt,
    roles: data.roles,
    dates: data.dates,
});

/**
 * Insert schema for the "public.users" table (after casing transform).
 */
export const UsersTableInsertSchema = z.pipe(UsersTableInsertBaseSchema, z.transform(transformUserInsertBaseRecord));

/**
 * Update schema for the "public.users" table (after casing transform).
 */
export const UsersTableUpdateSchema = z.pipe(z.partial(UsersTableInsertBaseSchema), z.transform(transformUserUpdateBaseRecord));

type TableInsertRecord = z.input<typeof UsersTableInsertSchema>;
type TableReadRecord = z.output<typeof UsersTableSchema>;

/**
 * Read record (casing transformed) for the "public.users" table.
 */
export interface UserRecord {
    /**
    */
    id: TableReadRecord['id'];
    /**
    */
    name: TableReadRecord['name'];
    /**
    */
    email?: TableReadRecord['email'];
    /**
    */
    status?: TableReadRecord['status'];
    /**
    */
    profile?: TableReadRecord['profile'];
    /**
    */
    createdAt?: TableReadRecord['createdAt'];
    /**
    */
    roles?: TableReadRecord['roles'];
    /**
    */
    dates?: TableReadRecord['dates'];
}

/**
 * Insert record (casing transformed) for the "public.users" table.
 */
export interface UserInsertRecord {
    /**
    * @maxLen: 100
    */
    name: TableInsertRecord['name'];
    /**
    * @maxLen: 255
    */
    email?: TableInsertRecord['email'];
    /**
    * @maxLen: 20
    * @default: 'active'::character varying
    */
    status?: TableInsertRecord['status'];
    /**
    */
    profile?: TableInsertRecord['profile'];
    /**
    * @default: now()
    */
    createdAt?: TableInsertRecord['createdAt'];
    /**
    */
    roles?: TableInsertRecord['roles'];
    /**
    */
    dates?: TableInsertRecord['dates'];
}

/**
 * Updatable record (casing transformed) for the "public.users" table.
 */
export type UserUpdateRecord = Partial<UserInsertRecord>;

export type UserRole = (typeof USER_ROLES)[number];
"
`;

exports[`generates schemas compatible with zod version 4 mini > types.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export type Table = 
    | 'users'
"
`;
